### PR TITLE
Validate config

### LINKED ISSUE #59: Validate Config
TODO: validate config
Commit: (1a0b16f) Update testing.
 [tests/MockStorage.js:85](https://github.com/digitalbazaar/edv-client/blob/master/tests/MockStorage.js#L85)
David I. Lehn commented a year ago

### PR DESCRIPTION
**Pull request by JSAssassin**
fixes #59

### CODE REVIEW

File: tests/ConfigSchema.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/*!
+ * Copyright (c) 2018-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+
+'use strict';
```
Comment (mattcollier) [inline tests/ConfigSchema.js:5]: no need for use strict in ES6 modules.  Please remove in all the places.

File: tests/ConfigSchema.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+export const edvConfig = {
+  title: 'EDV Configuration',
+  type: 'object',
+  // TODO: do not require primary `keyAgreementKey` and `hmac` in the future
```
Comment (mattcollier) [inline tests/ConfigSchema.js:10]: OH NO! A new TODO?  :) 
What is this about?  Is there an issue for it, can we resolve it now?

File: tests/validator.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+import {edvConfig} from './ConfigSchema';
+
+const Ajv = require('ajv');
```
Comment (davidlehn) [inline tests/validator.js:8]: Mixing module styles, should use `import`.

File: tests/ConfigSchema.js (hunk 2 above)
Comment (JSAssassin) [reply tests/ConfigSchema.js:10]: This was in the original one that I copied from. I wasn't sure if this is a change we want to make now...

Comment (dlongley) [reply tests/ConfigSchema.js:10]: `keyAgreementKey` and `hmac` should be optional in the config.

File: tests/validator.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+
```
Comment (dlongley) [inline tests/validator.js:4]:
```suggestion
```

File: tests/ConfigSchema.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
+/*!
+ * Copyright (c) 2018-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+
```
Comment (dlongley) [inline tests/ConfigSchema.js:4]:
```suggestion
```

File: tests/validator.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ajv.addSchema(edvConfig, 'edvConfig');
+
+// throws if validation fails
+export function validateSchema({payload}) {
+  // validate payload against JSON schema
+  const valid = ajv.validate('edvConfig', payload);
+  if(valid) {
+    return true;
+  }
+  const error = new SyntaxError('Validation error.');
```
Comment (davidlehn) [inline tests/validator.js:17]: Is `SyntaxError` the right choice here?  I think the specs say `SyntaxError` is for JavaScript parsing errors.  It's in test code so not a big deal.  Hard to map the few core errors to custom ones, but maybe just `Error` is better in this case?

File: tests/10-EdvClient.spec.js [hunk 7]
```diff
@@ -17,12 +16,18 @@ describe('EdvClient', () => {
...
     let result;
     let err;
     try {
-      result = validateSchema({payload: config});
+      result = await EdvClient.createEdv({
+        url: 'http://localhost:9876/edvs',
+        config: {
+          sequence: 0,
+          controller: mock.accountId,
+          invalid: 'invalid'
```
Comment (mattcollier) [inline tests/10-EdvClient.spec.js:28]: Please add a comment that you are intentionally adding an invalid property to the config.  The variable name is a good indicator, but I think it's best to be explicit when we're making negative tests so we're not wondering about what's going on.

**Pull request merged**